TTS/tts/models/__init__.py

```diff
@@ -4,20 +4,23 @@
 
 def setup_model(config):
     print(" > Using model: {}".format(config.model))
-
     MyModel = find_module("TTS.tts.models", config.model.lower())
     # define set of characters used by the model
     if config.characters is not None:
         # set characters from config
-        symbols, phonemes = make_symbols(**config.characters.to_dict())  # pylint: disable=redefined-outer-name
+        if hasattr(MyModel, "make_symbols"):
+            symbols = MyModel.make_symbols(config)
+        else:
+            symbols, phonemes = make_symbols(**config.characters)
     else:
         from TTS.tts.utils.text.symbols import phonemes, symbols  # pylint: disable=import-outside-toplevel
 
+        if config.use_phonemes:
+            symbols = phonemes
         # use default characters and assign them to config
         config.characters = parse_symbols()
-    num_chars = len(phonemes) if config.use_phonemes else len(symbols)
     # consider special `blank` character if `add_blank` is set True
-    num_chars = num_chars + getattr(config, "add_blank", False)
+    num_chars = len(symbols) + getattr(config, "add_blank", False)
```

Comment:
Maybe not the best place for this, but I was unable to run the pre-trained `ruslan` (Russian) or `baker` (Chinese) models with `v0.2.0`. Here are the errors:

ruslan:

```
 > Downloading model to /home/hansenm/.local/share/tts/tts_models--ru--ruslan--tacotron2-DDC
Traceback (most recent call last):
  File "/home/hansenm/opt/TTS/.venv/bin/tts", line 33, in <module>
    sys.exit(load_entry_point('TTS', 'console_scripts', 'tts')())
  File "/home/hansenm/opt/TTS/TTS/bin/synthesize.py", line 205, in main
    model_path, config_path, model_item = manager.download_model(args.model_name)
  File "/home/hansenm/opt/TTS/TTS/utils/manage.py", line 115, in download_model
    self._download_zip_file(model_item["github_rls_url"], output_path)
  File "/home/hansenm/opt/TTS/TTS/utils/manage.py", line 149, in _download_zip_file
    with zipfile.ZipFile(io.BytesIO(r.content)) as z:
  File "/usr/local/lib/python3.7/zipfile.py", line 1258, in __init__
    self._RealGetContents()
  File "/usr/local/lib/python3.7/zipfile.py", line 1325, in _RealGetContents
    raise BadZipFile("File is not a zip file")
zipfile.BadZipFile: File is not a zip file
```

---

baker:

```
Traceback (most recent call last):
  File "/home/hansenm/opt/TTS/.venv/bin/tts", line 33, in <module>
    sys.exit(load_entry_point('TTS', 'console_scripts', 'tts')())
  File "/home/hansenm/opt/TTS/TTS/bin/synthesize.py", line 206, in main
    args.vocoder_name = model_item["default_vocoder"] if args.vocoder_name is None else args.vocoder_name
KeyError: 'default_vocoder'
```

---

Below is the script I used to check all of the models available with `--list_models`. Running it in the root of `TTS` with a directory name as an argument will create a WAV file for each model as well as some benchmarking info.

```bash
#!/usr/bin/env bash
if [[ -z "$1" ]]; then
    echo "Usage: make_voice_samples.sh <OUTPUT_DIR> [MODEL]..."
    exit 1
fi

output_dir="$1"
shift 1  # Remaining arguments will optionally be model names

mkdir -p "${output_dir}"

# Directory of *this* script
this_dir="$( cd "$( dirname "$0" )" && pwd )"

# Avoid bash built-in "time" command that doesn't allow output to a file
time_prog="$(which time)"

# -----------------------------------------------------------------------------

# Pangrams from: http://clagnut.com/blog/2380/
declare -A pangrams
pangrams['en']='The beige hue on the waters of the loch impressed all, including the French queen, before she heard that symphony again, just as young Arthur wanted.'
pangrams['es']='Benjamín pidió una bebida de kiwi y fresa; Noé, sin vergüenza, la más exquisita champaña del menú.'
pangrams['de']='Falsches Üben von Xylophonmusik quält jeden größeren Zwerg.'
pangrams['fr']='Buvez de ce whisky que le patron juge fameux.'
pangrams['it']='Ma la volpe, col suo balzo, ha raggiunto il quieto Fido.'
pangrams['ja']='色は匂へど 散りぬるを 我が世誰ぞ 常ならむ 有為の奥山 今日越えて 浅き夢見じ 酔ひもせず'
pangrams['nl']='Pa’s wijze lynx bezag vroom het fikse aquaduct.'
pangrams['ru']='Широкая электрификация южных губерний даст мощный толчок подъёму сельского хозяйства.'
pangrams['sv']='Yxskaftbud, ge vår WC-zonmö IQ-hjälp.'
pangrams['zh-CN']='視野無限廣，窗外有藍天，微風迎客，軟語伴茶'

# -----------------------------------------------------------------------------

# Speaker indexes for multi-speaker models
declare -A speaker_idxs
speaker_idxs['tts_models/en/vctk/sc-glow-tts']='p225'

function test_model {
    model_path="$1"

    # Split on / into tts_models/<lang>/<dataset>/<model_type>
    IFS='/' read -ra name_parts <<< "${model_path}"
    model_lang="${name_parts[1]}"
    model_name="${name_parts[2]}"
    model_type="${name_parts[3]}"

    # Get pangram for this language
    pangram="${pangrams["${model_lang}"]}"
    if [[ -z "${pangram}" ]]; then
        echo "WARNING: No pangram available for language ${model_lang}"
        continue
    fi

    # Extra arguments for multi-speaker models
    tts_extra_args=()
    speaker_idx="${speaker_idxs["${model_path}"]}"
    if [[ -n "${speaker_idx}" ]]; then
        tts_extra_args+=('--speaker_idx' "${speaker_idx}")
    fi

    # Write WAV and two text files, one with timing info and the other with stdout/stderr
    sample_prefix="${model_lang}--${model_name}--${model_type}"
    sample_wav="${output_dir}/${sample_prefix}.wav"
    sample_time="${output_dir}/${sample_prefix}_time.txt"
    sample_error="${output_dir}/${sample_prefix}_output.txt"

    # Time TTS command
    echo "${model_path}"
    "${time_prog}" "--output=${sample_time}" -- \
                   tts --model_name "${model_path}" --text "${pangram}" --out_path "${sample_wav}" "${tts_extra_args[@]}" 2>&1 \
        | tee "${sample_error}"

}

# -----------------------------------------------------------------------------

if [[ -z "$1" ]]; then
    # All models from --list_models
    while read -r model_path; do
        test_model "${model_path}"
    done < <(tts --list_models | grep 'tts_models/' | sed -e 's/^[^t]\+//' | cut -d' ' -f1)
else
    # Specific model(s)
    while [[ -n "$1" ]]; do
        test_model "$1"
        shift 1
    done
fi
```

Reply:
thanks for the 🐛 catch. 

Baker model should be fixed now. Ruslan has been deleted due to the license conflict but I forgot to remove it from the list. Now it is gone :( 

The script looks amazing.  I copied to use it in local tests

Reply:
> This causes most of the pre-trained models to fail to load because `symbols` is not set to `phonemes` if `config.characters` is `None`. In this case, `num_chars` incorrectly reflects the number of symbols instead of phonemes.

you might overlook line 19. I think it fixes the issue you mentioned.

Reply:
@

> Regarding the speaker manager, I had issues with Thorsten's Tacotron2 not loading. I think the [problem is here on line 35 of tacotron2.py](https://github.com/coqui-ai/TTS/blob/d0292dd2d1dfafbd68290965929f679d10959be6/TTS/tts/models/tacotron2.py#L35):
> 
> ```python
>         if self.num_speakers > 1:
>             self.init_multispeaker(config)
> ```
> 
> With a single speaker model, `init_multispeaker` won't be called and so the `speaker_manager` attribute [won't be set in the class](https://github.com/coqui-ai/TTS/blob/d0292dd2d1dfafbd68290965929f679d10959be6/TTS/tts/models/base_tts.py#L61). Downstream in `server.py` then, accessing the TTS model's `speaker_manager` directly throws an exception.
> 
> One way to solve this would be to add in the `__init__` of `BaseTTS`:
> 
> ```python
> self.speaker_manager: Optional[SpeakerManager] = None
> ```

I've made the change you suggested above.